firedrake/slate/slac/compiler.py

```diff
@@ -530,6 +590,325 @@ def parenthesize(arg, prec=None, parent=None):
+            from pymbolic.mapper.stringifier import PREC_NONE
+            from loopy.symbolic import SubArrayRef
+            from pymbolic import var
+
+            mat_descr = self.arg_id_to_descr[0]
+
+            arg_c_parameters = [
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                if isinstance(par, SubArrayRef) else
+                expression_to_code_mapper(
+                    par,
+                    PREC_NONE,
+                    dtype_to_type_context(target, par_dtype),
+                    par_dtype
+                ).expr
+                for par, par_dtype in zip(parameters, par_dtypes)
+            ]
+            c_parameters = []
+            c_parameters.insert(0, arg_c_parameters[0])  # t1
+            c_parameters.insert(1, mat_descr.shape[0])  # n
+            return var(self.name_in_target)(*c_parameters), False
+
+        def generate_preambles(self, target):
+            assert isinstance(target, CTarget)
+            inverse_preamble = """
+                #include <string.h>
+                #include <stdio.h>
+                #include <stdlib.h>
+                #ifndef Inverse_HPP
+                #define Inverse_HPP
+                void inverse_(PetscScalar* A, PetscBLASInt N)
+                {
+                    PetscBLASInt info;
+                    PetscBLASInt* ipiv=(PetscBLASInt*) malloc(N*sizeof(PetscBLASInt));
+                    PetscScalar* Awork=(PetscScalar*) malloc(N*N*sizeof(PetscScalar));
+                    LAPACKgetrf_(&N,&N,A,&N,ipiv,&info);
+                    if(info==0)
+                        LAPACKgetri_(&N,A,&N,ipiv,Awork,&N,&info);
+                    if(info!=0)
+                        fprintf(stderr,\"Getri throws nonzero info.\");
+                }
+                #endif
+            """
+            yield("lapack_inverse", "#include <petscsystypes.h>\n#include <petscblaslapack.h>\n"+inverse_preamble)
+            return
+
+    def inv_fn_lookup(target, identifier):
+        if identifier == 'inv':
+            return INVCallable(name='inv')
+
+        return None
+
+    loopy_merged = loopy.register_function_id_to_in_knl_callable_mapper(loopy_merged, inv_fn_lookup)
+
+    return loopy_merged
+
+
+def get_solve_callable(loopy_merged):
+    class SolveCallable(loopy.ScalarCallable):
+        def __init__(self, name, arg_id_to_dtype=None, 
+                     arg_id_to_descr=None, name_in_target=None):
+
+            super(SolveCallable, self).__init__(name,
+                  arg_id_to_dtype=arg_id_to_dtype,
+                  arg_id_to_descr=arg_id_to_descr)
+
+            self.name = name
+            self.name_in_target = name_in_target
+            
+        def with_types(self, arg_id_to_dtype, kernel, callables_table):
+            for i in range(0, len(arg_id_to_dtype)):
```

Comment:
`range` defaults to starting from 0, so you could just write `range(len(arg_id_to_dtype))`